Collection items map

rowanwins opened the pull request:
This PR adds support for viewing Collection Item extents on the preview map. And when [this PR](https://github.com/stac-utils/stac-layer/pull/34) get's merged in `stac-layer` it'll also show the thumbnail or overview for each item on the map.

In theory the extents and thumbnails update when the user uses the pagination, although there does seem to be some sort of bug in removing the old items when a user pages, I think this is coming from upstream in georaster-for-leaflet, so I'm currently tracking that down.

https://user-images.githubusercontent.com/6735870/163709004-732f50ef-7ec1-40a0-8349-83ad3de9bd95.mp4

I suggest we keep this in draft until I can work out what's going on upstream.

m-mohr commented on src/views/Catalog.vue:
```diff
@@ -160,7 +160,7 @@ export default {
       return {
         type: 'FeatureCollection',
-        features: this.items
+        features: this.items.map(i => i._rawJson)
```
Is this actually required? The STAC class is a decorator to the STAC object, so I'd assume you can simply pass it over.

Another side note: To keep the code base consistent, I'd like to see `;` at the end of the lines. :-) Actually, I thought eslint would ask for them, but maybe I need to add that.

rowanwins replied:
I added this when having problems with clearing the stachlayer, I thought perhaps the class object might be having some side effects, so happy to remove it.
And yep add an eslint rule and I'll follow it 🙂

m-mohr replied:
Ah, okay, then we should investigate the side effects and try to avoid those :-) Any details?

m-mohr replied:
I've added eslint rules in the v3 branch, it should fix most issues automatically once you run `npm run lint`.

rowanwins replied:
Have addressed comment. And the weirdness with the stac-layer dependency seems to have disappeared, maybe because versions got pinned somewhere.

m-mohr commented:
This seems to work quite good already, thanks.

Are there already precautions to limit the number of COGs loaded once https://github.com/stac-utils/stac-layer/pull/34 gets merged and released?

One thing I don't like right now is that is zooms to the item extents. The primary use case of the box right now is to show the extent of the collection, but it is not directly visible when zooming in too far.

I'm also wondering whether this should also show item extents for static catalogs? Right now the PR only works for APIs, it seems.

m-mohr commented:
This looks better now, but I still get a weird zooming behavior for planetarycomputer.microsoft.com/api/stac/v1/collections/gnatsgo-tables

![grafik](https://user-images.githubusercontent.com/8262166/175833490-498cc7bf-e408-4ddb-8b05-5cfc67cc6b66.png)

Also, there are some where it looks really weird if they are stacked over each other. The example above shows it also. More examples:
- https://planetarycomputer.microsoft.com/api/stac/v1/collections/chloris-biomass
- https://planetarycomputer.microsoft.com/api/stac/v1/collections/eclipse
- https://planetarycomputer.microsoft.com/api/stac/v1/collections/deltares-floods

And then I found a collection, where it looks a bit weird and from a user perspective it could be confusing to see the two boxes with one being thicker, inside and adding some "diagonal" lines: https://planetarycomputer.microsoft.com/api/stac/v1/collections/mobi

![grafik](https://user-images.githubusercontent.com/8262166/175833480-099953ef-f7b5-4d69-a697-1823c69c69be.png)

I think we should tune the styling a bit more, e.g. the border for the items is bigger as the overall bounding box.
I'm not sure yet how to improve this, but an idea could be to the  make the border as slim as the overall extent (1px?), have a different color and maybe not have a fill?

rowanwins commented:
> This looks better now, but I still get a weird zooming behavior for planetarycomputer.microsoft.com/api/stac/v1/collections/gnatsgo-tables

The `extent.spatial.bbox` property on this STAC collection is pretty wild, it's got 9 bbox arrays! I've fixed it up so we now take all those into account. In the currently deployed version you see the 9 bounding boxes, although to the user it probably wouldn't be entirely clear what they were looking at. 

> Also, there are some where it looks really weird if they are stacked over each other. The example above shows it also. More examples:

The items in the gnatso-tables collection largely seem to share a common bbox and geometry (at least the first 3 pages that I looked at). I'm not terribly sure what the best option is here... having no fill means you just see a single outline which is probably an improvement, but it also doesn't give the user much value...


> I think we should tune the styling a bit more, e.g. the border for the items is bigger as the overall bounding box.
I'm not sure yet how to improve this, but an idea could be to the make the border as slim as the overall extent (1px?), have a different color and maybe not have a fill?

stac-layer doesn't provide an option for styling so the workaround for the timebeing would involve using leafletjs [eachLayer](https://leafletjs.com/reference.html#layergroup-eachlayer) function with some sort of nesting and applying styling based on that, it would be a little clunky but we could make it work.
Alternatively if all items had the same extent as the collection then don't show the item extents?

> And then I found a collection, where it looks a bit weird and from a user perspective it could be confusing to see the two boxes with one being thicker, inside and adding some "diagonal" lines:

I've added some styling on the bbox we're adding so we're at least controlling what it easily within our grasp.

m-mohr commented:
> The `extent.spatial.bbox` property on this STAC collection is pretty wild, it's got 9 bbox arrays! I've fixed it up so we now take all those into account. In the currently deployed version you see the 9 bounding boxes, although to the user it probably wouldn't be entirely clear what they were looking at.

Actually, this is another reason to implement this in stac-layer and not here, because otherwise we are duplicating it. Here the implementation is actually wrong. while in STAC layer it works correctly. They way it is expected to work in STAC is that if you have a single bbox, then show it, otherwise if you have multiple bboxes, you should show all of them (except the first one, which is the union of the other ones).

> stac-layer doesn't provide an option for styling so the workaround for the timebeing would involve using leafletjs [eachLayer](https://leafletjs.com/reference.html#layergroup-eachlayer) function with some sort of nesting and applying styling based on that, it would be a little clunky but we could make it work. Alternatively if all items had the same extent as the collection then don't show the item extents?

Let's add this to STAC Layer then. I'll check this now...
